src/java/com/unifina/signalpath/utils/MessageChainUtil.java

```diff
@@ -0,0 +1,73 @@
+package com.unifina.signalpath.utils;
+
+import com.streamr.client.protocol.message_layer.MessageID;
+import com.streamr.client.protocol.message_layer.MessageRef;
+import com.streamr.client.protocol.message_layer.StreamMessage;
+import com.streamr.client.protocol.message_layer.StreamMessageV30;
+import com.unifina.data.StreamPartitioner;
+import com.unifina.domain.data.Stream;
+import com.unifina.domain.security.SecUser;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MessageChainUtil implements Serializable {
+	private Map<String,Long> previousTimestamps = new HashMap<>();
+	private Map<String,Long> previousSequenceNumbers = new HashMap<>();
+
+	private transient SecUser cachedUser = null;
+
+	private static final Logger log = Logger.getLogger(MessageChainUtil.class);
+
+	private SecUser getUser(Long userId) {
+		if (cachedUser == null) {
+			cachedUser = SecUser.getViaJava(userId);
+		}
+		return cachedUser;
+	}
+
+	private long getNextSequenceNumber(String key, long timestamp) {
+		if (!previousTimestamps.containsKey(key) || previousTimestamps.get(key) != timestamp) {
+			return 0L;
+		}
+		if (!previousSequenceNumbers.containsKey(key)) {
```

Comment:
Optimization: 1 hashmap lookup instead of 2:
```
Long previousSequenceNumber = previousSequenceNumbers.get(key);
return previousSequenceNumber == null ? 0 : previousSequenceNumber + 1;
```
(In this suggestion I also removed the `put`, because `getStreamMessage` does it anyway. Also better pattern not to write anything in methods called `getXXXX`)